test/clj/rems/test_browser.clj

```diff
@@ -625,3 +625,108 @@
       (login-as "alice")
       (go-to-catalogue)
       (btu/visible? {:fn/text (btu/context-get :catalogue-item-name)}))))
+
+(deftest test-form-editor
+  (btu/with-postmortem {:dir btu/reporting-dir}
+    (login-as "owner")
+    (go-to-admin-forms)
+    (btu/scroll-and-click :create-form)
+    (btu/wait-visible {:tag :h1 :fn/text "Create form"})
+    (select-option "Organization" "nbn")
+    (fill-form-field "Form name" "Form editor test")
+    (btu/scroll-and-click {:class :add-form-field})
+    ;; using ids to fill the fields because the label structure is complicated
+    (btu/wait-visible :fields-0-title-en)
+    (btu/fill-human :fields-0-title-en "Text area (EN)")
+    (btu/fill-human :fields-0-title-fi "Text area (FI)")
+    (btu/fill-human :fields-0-title-sv "Text area (SV)")
+    (btu/fill-human :fields-0-placeholder-en "Placeholder (EN)")
+    (btu/fill-human :fields-0-placeholder-fi "Placeholder (FI)")
+    (btu/fill-human :fields-0-placeholder-sv "Placeholder (SV)")
+    (btu/scroll-and-click :fields-0-type-texta)
+    (btu/scroll-and-click :fields-0-optional)
+    (btu/fill-human :fields-0-max-length "127")
+
+    (btu/scroll-and-click-el (last (btu/query-all {:class :add-form-field})))
+    (btu/wait-visible :fields-1-title-en)
+    (btu/fill-human :fields-1-title-en "Option list (EN)")
+    (btu/fill-human :fields-1-title-fi "Option list (FI)")
+    (btu/fill-human :fields-1-title-sv "Option list (SV)")
+    (btu/scroll-and-click :fields-1-type-option)
+    (btu/scroll-and-click {:class :add-option})
+    (btu/wait-visible :fields-1-options-0-key)
+    (btu/fill-human :fields-1-options-0-key "true")
+    (btu/fill-human :fields-1-options-0-label-en "Yes")
+    (btu/fill-human :fields-1-options-0-label-fi "Kyllä")
+    (btu/fill-human :fields-1-options-0-label-sv "Ja")
+    (btu/scroll-and-click {:class :add-option})
+    (btu/wait-visible :fields-1-options-1-key)
+    (btu/fill-human :fields-1-options-1-key "false")
+    (btu/fill-human :fields-1-options-1-label-en "No")
+    (btu/fill-human :fields-1-options-1-label-fi "Ei")
+    (btu/fill-human :fields-1-options-1-label-sv "Nej")
+
+    ;; TODO create all field types?
+    ;; TODO test validations?
+
+    (btu/scroll-and-click :save)
+    (btu/wait-visible {:tag :h1 :fn/text "Form"})
+    (btu/wait-page-loaded)
+
+    (is (= {"Organization" "NBN"
+            "Title" "Form editor test"
+            "Active" ""}
+           (slurp-fields :form)))
+    ;; the text is split into multiple DOM nodes so we need to get fancy
```

Comment:
I think this is something that should be fixed. We could for example have a structured localized text for title so that maxlength and required are just parameters. Or since this part mostly does text manipulation wrap it into a `str` but that would prevent using elements in the parts.

Reply:
Perhaps you could put here an explanation that this is checking the preview?

Reply:
I'll add an explanation.

I would rather implement a generic helper that loops through elements and uses get-element-text: otherwise we need to jump through hoops in the implementation every time

Reply:
Maybe extract this part to its own localized text?
```
      title " "
      (when max-length
        (text-format :t.form/maxlength (str max-length)))
      " "
      (if optional
        (text :t.form/optional)
        (text :t.form/required))]
```

Reply:
Ideally `:fn/has-text` would work the same as `has-text?` in etaoin.

Reply:
Looks like this would be possible in xpath and `has-text?` uses internally `xpath/node-by-text` whereas `fn/has-text` uses a simpler xpath that does not "do the right thing". Could try patching this.

Reply:
I've fixed some sites in the `application.cljs` previously to make the DOM simpler.

Reply:
I'll see if `has-text?` works